### PR TITLE
Implement 'join' for optional types and a special case for 'nil'.

### PR DESCRIPTION
**Pull request by DougGregor**
#### What's in this pull request?

The join operation for optional types is straightforward to define/implement:

```
join(T?, U)  ::= join(T, U)?
join(T, U?)  ::= join(T, U)?
join(T?, U?) ::= join(T, U)?
```

As a special case in the constraint solver, handle the join of a 'nil'
literal with a non-ExpressibleByNilLiteral-conforming concrete type
'T' to produce 'T?'. This allows us, e.g., infer [String?] for the
expressions

```
["hello", nil]
```

and

```
 true ? "hello" nil
```

for example.
#### Resolved bug number: ([rdar://problem/16326914](rdar://problem/16326914))

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| All supported platforms | @swift-ci Please smoke test and merge |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

A smoke test on macOS does the following:
1. Builds the compiler incrementally.
2. Builds the standard library only for macOS. Simulator standard libraries and
   device standard libraries are not built.
3. lldb is not built.
4. The test and validation-test targets are run only for macOS. The optimized
   version of these tests are not run.

A smoke test on Linux does the following:
1. Builds the compiler incrementally.
2. Builds the standard library incrementally.
3. lldb is built incrementally.
4. The swift test and validation-test targets are run. The optimized version of these
   tests are not run.
5. lldb is tested.

**Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| All supported platforms | @swift-ci Please test and merge |
| OS X platform | @swift-ci Please test OS X platform |
| OS X platform | @swift-ci Please benchmark |
| Linux platform | @swift-ci Please test Linux platform |

**Lint Testing**

| Language | Comment |
| --- | --- |
| Python | @swift-ci Please Python lint |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please test macOS

File: lib/AST/TypeJoinMeet.cpp [hunk 1]
```diff
@@ -1,4 +1,4 @@
-//===--- TypeJoinMeet.cpp - Swift Type "Join" and "Meet"  -----------------===//
+//===--- TypeJoinjoin.cpp - Swift Type "Join" and "join"  -----------------===//
```
Comment (jrose-apple) [inline lib/AST/TypeJoinMeet.cpp]: sed gone wild, heh.

File: lib/AST/TypeJoinMeet.cpp [hunk 2]
```diff
@@ -10,7 +10,7 @@
 //
 //===----------------------------------------------------------------------===//
 //
-//  This file implements the "meet" operation for types (and, eventually,
+//  This file implements the "join" operation for types (and, eventually,
```
Comment (jrose-apple) [inline lib/AST/TypeJoinMeet.cpp]: and here

File: lib/AST/TypeJoinMeet.cpp (hunk 1 above)
Comment (jckarter) [inline lib/AST/TypeJoinMeet.cpp]: Doug is hereby enmeeted from using `sed` again.

**Comment by DougGregor** [comment]
Drat! Will fix.

**Comment by jrose-apple** [comment]
I'm not a type checker expert but everything else looks good.

**Comment by DougGregor** [comment]
This isn't specifically addressing any regression that I know of (I originally thought it did), so probably not for 3.0. Whaddya think, @rudkx ?

**Comment by DougGregor** [comment]
Force-merging; the CI shouldn't pay for my comment fixes.

**Comment by rudkx** [comment]
@DougGregor Yeah, I don't think we need to take this now.

**Pull request merged**